source/Octopus.Server.Client/Exceptions/OctopusExceptionFactory.cs

```diff
@@ -51,74 +51,96 @@ public static async Task<OctopusException> CreateException(HttpResponseMessage r
+                return new OctopusMethodNotAllowedFoundException(OctopusErrorsContractFromBody(body).ErrorMessage);
             }
 
-            var fullDetails = body;
-            string helpText = null;
-            try
+            return CreateOctopusServerException(statusCode, body);
+        }
+
+        private static OctopusSecurityException CreateOctopusSecurityException(int statusCode, string body)
+        {
+            var errorsContract = OctopusErrorsContractFromBody(body);
+
+            return new OctopusSecurityException(statusCode, errorsContract.ErrorMessage)
             {
-                var errors = JsonConvert.DeserializeObject<OctopusErrorsContract>(body);
-                if (errors != null)
-                {
-                    fullDetails = "Octopus Server returned an error: " + errors.ErrorMessage;
-                    helpText = errors.HelpText;
-                    if (!string.IsNullOrWhiteSpace(errors.FullException))
-                    {
-                        fullDetails += Environment.NewLine + "Server exception: " + Environment.NewLine + errors.FullException + Environment.NewLine + "-----------------------" + Environment.NewLine;
-                    }
-                }
-            }
-            // ReSharper disable once EmptyGeneralCatchClause
-            catch
+                HelpText = errorsContract.HelpText
+            };
+        }
+
+        private static OctopusValidationException CreateOctopusValidationException(int statusCode, string body)
+        {
+            var errorsContract = OctopusErrorsContractFromBody(body);
+
+            return new OctopusValidationException(
+                statusCode,
+                errorsContract.ErrorMessage,
+                errorsContract.Errors,
+                errorsContract.Details)
+            {
+                HelpText = errorsContract.HelpText
+            };
+        }
+
+        private static OctopusServerException CreateOctopusServerException(int statusCode, string body)
+        {
+            var errorsContract = OctopusErrorsContractFromBody(body);
+            var fullDetails = errorsContract.ErrorMessage;
+            if (!string.IsNullOrWhiteSpace(errorsContract.FullException))
             {
+                fullDetails =
+                    $"Octopus Server returned an error: {errorsContract.ErrorMessage} {Environment.NewLine}Server exception: {Environment.NewLine}{errorsContract.FullException}{Environment.NewLine} ----------------------- {Environment.NewLine}";
             }
 
-            return new OctopusServerException(statusCode, fullDetails) {HelpText = helpText};
+            return new OctopusServerException(statusCode, fullDetails)
+            {
+                HelpText = errorsContract.HelpText
+            };
         }
 
-        static string GetErrorMessage(string body)
+        internal static OctopusErrorsContract OctopusErrorsContractFromBody(string body)
```

Comment:
`internal` to support `expected result` for unit tests

Reply:
This can be made private again.